auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt

```diff
@@ -0,0 +1,241 @@
+package uk.gov.communities.delta.auth.controllers.internal
+
+import io.ktor.http.*
+import io.ktor.server.application.*
+import io.ktor.server.auth.*
+import io.ktor.server.request.*
+import io.ktor.server.response.*
+import io.ktor.server.routing.*
+import kotlinx.serialization.Serializable
+import org.slf4j.LoggerFactory
+import uk.gov.communities.delta.auth.config.LDAPConfig
+import uk.gov.communities.delta.auth.plugins.ApiError
+import uk.gov.communities.delta.auth.repositories.LdapUser
+import uk.gov.communities.delta.auth.repositories.isInternal
+import uk.gov.communities.delta.auth.services.*
+
+class EditAccessGroupsController(
+    private val userLookupService: UserLookupService,
+    private val groupService: GroupService,
+    private val organisationService: OrganisationService,
+    private val accessGroupsService: AccessGroupsService,
+    private val memberOfToDeltaRolesMapperFactory: MemberOfToDeltaRolesMapperFactory,
+) {
+
+    private val logger = LoggerFactory.getLogger(javaClass)
+
+    fun route(route: Route) {
+        route.post { updateUserAccessGroups(call) }
+    }
+
+    // This endpoint takes a list of access groups to add and to remove in the form of maps of the group name (with
+    // delta prefix) and a list of organisations for that access group. Generation of the correct lists is performed in
+    // Delta.
+    private suspend fun updateUserAccessGroups(call: ApplicationCall) {
+        val session = call.principal<OAuthSession>()!!
+        val callingUser = userLookupService.lookupUserByCn(session.userCn)
+        logger.atInfo().log("Updating access groups for user {}", session.userCn)
+        val userIsInternal = callingUser.isInternal()
+        val allAccessGroups = accessGroupsService.getAllAccessGroups().associateBy { it.name }
+
+        val requestBodyObject = call.receive<DeltaUserAccessGroups>()
+        val selectedOrgs = requestBodyObject.userSelectedOrgs.toSet()
+        val accessGroupsRequestMap = stripDatamartPrefixFromKeys(requestBodyObject.accessGroupsRequest)
+        val accessGroupsRequested = accessGroupsRequestMap.filter { m -> m.value.isNotEmpty() }
+
+        validateRequest(accessGroupsRequested, selectedOrgs, allAccessGroups, userIsInternal, callingUser)
```

Comment:
We should also check the access groups we're removing users from are actual access groups as well.

Reply:
Is there more to that permission than the existing organisation checks?